cirq/circuits/circuit_test.py

```diff
@@ -581,44 +579,126 @@ def test_concatenate_with_device():
     assert len(cone) == 0
 
 
+class FakeXmonDevice(cirq.Device):
+    """A Fake XMon device that was created to ensure certain Device validation features are
+    leveraged in Circuit functions. Originally this was cirq.google.Foxtail and
+    cirq.google.Bristlecone, however we decoupled cirq.google from cirq core. This class now
+    contains the minimum set of features that tests required. Feel free to extend the features
+    here as needed."""
+
+    def __init__(
+        self,
+        allowed_qubit_types: Tuple[type, ...],
+        allowed_gates: Tuple[type, ...],
+        qubits: AbstractSet[cirq.Qid],
+        name: str,
+    ):
+        self.allowed_qubit_types = allowed_qubit_types
+        self.allowed_gates = allowed_gates
+        self.qubits = qubits
+        self._repr = name
+
+    def validate_operation(self, operation: cirq.Operation) -> None:
+        # This is pretty close to what the cirq.google.XmonDevice has for validation
+        for q in operation.qubits:
+            if not isinstance(q, self.allowed_qubit_types):
+                raise ValueError("Unsupported qubit type: {!r}".format(type(q)))
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+        if not isinstance(operation.gate, self.allowed_gates):
+            raise ValueError("Unsupported gate type: {!r}".format(operation.gate))
+        if len(operation.qubits) == 2 and not isinstance(operation.gate, ops.MeasurementGate):
+            p, q = operation.qubits
+            if not cast(cirq.GridQubit, p).is_adjacent(q):
+                raise ValueError('Non-local interaction: {!r}.'.format(operation))
+
+    def decompose_operation(self, operation: 'cirq.Operation') -> 'cirq.OP_TREE':
+        # a fake decomposer for only TOFFOLI gates
+        if isinstance(operation.gate, cirq.CCXPowGate):
+            return cirq.decompose(operation)
+        return operation
+
+    def can_add_operation_into_moment(
+        self, operation: 'cirq.Operation', moment: 'cirq.Moment'
+    ) -> bool:
+        # a fake rule for ensuring that no two CZs are executed at the same moment.
+        # this will ensure that CZs are always in separate moments in this device
+        return not (
+            isinstance(operation.gate, cirq.CZPowGate)
+            and any(isinstance(op.gate, cirq.CZPowGate) for op in moment.operations)
+        )
+
+    def __repr__(self):
+        return self._repr
+
+
+foxy = FakeXmonDevice(
+    allowed_qubit_types=(cirq.GridQubit,),
+    allowed_gates=(
+        ops.CZPowGate,
+        ops.XPowGate,
+        ops.YPowGate,
+        ops.ZPowGate,
+    ),
+    qubits={
+        cirq.GridQubit(0, 0),
+        cirq.GridQubit(1, 0),
+        cirq.GridQubit(0, 1),
+        cirq.GridQubit(1, 1),
+        cirq.GridQubit(0, 6),
```

Comment:
This may generate confusion - I'd recommend `cirq.GridQubit.rect(2,7)` instead.

Reply:
done.